synapse/util/caches/lrucache.py

```diff
@@ -484,3 +750,8 @@ def set_cache_factor(self, factor: float) -> bool:
                 self._on_resize()
             return True
         return False
+
+    def __del__(self) -> None:
+        # We're about to be deleted, so we make sure to clear up all the nodes
+        # and run callbacks, etc.
+        self.clear()
```

Comment:
not saying this is a bad thing, but what is the thinking here? Do we expect LruCache objects to get deleted in operation, or is this just for tests?

Reply:
OMG this took ages to track down, but in the sync code we have an expiring cache of lru caches: 

https://github.com/matrix-org/synapse/blob/fcf3c7032b96ab454120f86f3f070160c409d599/synapse/handlers/sync.py#L280-L286

which led to all sorts of fun and games with orphaned nodes.